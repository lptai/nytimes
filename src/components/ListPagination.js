import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Flex } from 'grid-styled';

const PageItem = styled.li.attrs({
  className: 'page-item',
})`
  display: inline;
  & > a {
    position: relative;
    float: left;
    padding: 6px 12px;
    line-height: 1.42857143;
    text-decoration: none;
    color: ${props => (props.isCurrentPage ? 'white' : '#337ab7')};
    background-color: ${props => (props.isCurrentPage ? '#337ab7' : '#fff')};
    border: 1px solid #ddd;
    margin-left: -1px;
    &:focus,
    &:hover {
    ${props =>
      props.isCurrentPage
        ? `
        z-index: 3;
        color: #fff;
        background-color: #337ab7;
        border-color: #337ab7;
        cursor: default;`
        : `
      z-index: 2;
      color: #23527c;
      background-color: #eee;
      border-color: #ddd;
    }`};
  }
`;

const ListPagination = props => {
  const { totalArticle, maxItemsPerPage, currentPage, onPageChange } = props;

  if (totalArticle < maxItemsPerPage) {
    return null;
  }

  const pages = [];

  for (let i = 0; i < totalArticle / maxItemsPerPage; i++) {
    pages.push(i);
  }

  const buildPages = () =>
    pages.map((page, idx) => {
      const isCurrentPage = page === currentPage;

      const onPageClick = e => {
        e.preventDefault();
        if (isCurrentPage) {
          return;
        }
        onPageChange(page);
      };
      return (
        <PageItem key={idx} className="" onClick={onPageClick} isCurrentPage={isCurrentPage}>
          <a className="page-link" href="#">
            {page + 1}
          </a>
        </PageItem>
      );
    });

  return (
    <div className="pagination">
      <ul className="page">
        <Flex flexDirection={'row'} justifyContent={'center'}>
          {buildPages()}
        </Flex>
      </ul>
    </div>
  );
};

ListPagination.propTypes = {
  maxItemsPerPage: PropTypes.number,
  totalArticle: PropTypes.number,
};

ListPagination.defaultProps = {
  maxItemsPerPage: 10,
  totalArticle: 100,
};

export { PageItem };

export default ListPagination;
