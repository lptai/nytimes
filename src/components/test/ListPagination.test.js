import React from 'react';

import 'jest-styled-components';

import Enzyme, { shallow } from 'enzyme';
import ListPagination, { PageItem } from './../ListPagination';

test('Pagination test > Expect generate 10 page items', () => {
  const pagination = shallow(
    <ListPagination currentPage={0} maxItemsPerPage={10} totalArticle={200} />,
  );
  expect(pagination.find(PageItem).length).toBe(20);
  expect(pagination).toMatchSnapshot();

  const emptyPagination = shallow(
    <ListPagination currentPage={0} maxItemsPerPage={10} totalArticle={0} />,
  );

  expect(emptyPagination.find(PageItem).length).toBe(0);
  expect(emptyPagination).toMatchSnapshot();

  const defaultPagination = shallow(<ListPagination currentPage={0} />);

  expect(defaultPagination.find(PageItem).length).toBe(10);
  expect(defaultPagination).toMatchSnapshot();
});
