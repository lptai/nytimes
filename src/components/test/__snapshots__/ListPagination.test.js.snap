// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test > Expect generate 10 page items 1`] = `
<div
  className="pagination"
>
  <ul
    className="page"
  >
    <Flex
      flexDirection="row"
      justifyContent="center"
      theme={
        Object {
          "breakpoints": Array [
            "40em",
            "52em",
            "64em",
          ],
          "fontSizes": Array [
            12,
            14,
            16,
            20,
            24,
            32,
            48,
            64,
            96,
            128,
          ],
          "space": Array [
            0,
            4,
            8,
            16,
            32,
            64,
            128,
            256,
            512,
          ],
        }
      }
    >
      <styled.li
        isCurrentPage={true}
        key="0"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          1
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="1"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          2
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="2"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          3
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="3"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          4
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="4"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          5
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="5"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          6
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="6"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          7
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="7"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          8
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="8"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          9
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="9"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          10
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="10"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          11
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="11"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          12
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="12"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          13
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="13"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          14
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="14"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          15
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="15"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          16
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="16"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          17
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="17"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          18
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="18"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          19
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="19"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          20
        </a>
      </styled.li>
    </Flex>
  </ul>
</div>
`;

exports[`Pagination test > Expect generate 10 page items 2`] = `""`;

exports[`Pagination test > Expect generate 10 page items 3`] = `
<div
  className="pagination"
>
  <ul
    className="page"
  >
    <Flex
      flexDirection="row"
      justifyContent="center"
      theme={
        Object {
          "breakpoints": Array [
            "40em",
            "52em",
            "64em",
          ],
          "fontSizes": Array [
            12,
            14,
            16,
            20,
            24,
            32,
            48,
            64,
            96,
            128,
          ],
          "space": Array [
            0,
            4,
            8,
            16,
            32,
            64,
            128,
            256,
            512,
          ],
        }
      }
    >
      <styled.li
        isCurrentPage={true}
        key="0"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          1
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="1"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          2
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="2"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          3
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="3"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          4
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="4"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          5
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="5"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          6
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="6"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          7
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="7"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          8
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="8"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          9
        </a>
      </styled.li>
      <styled.li
        isCurrentPage={false}
        key="9"
        onClick={[Function]}
      >
        <a
          className="page-link"
        >
          10
        </a>
      </styled.li>
    </Flex>
  </ul>
</div>
`;
